src/rgw/rgw_tracer.cc

```diff
@@ -0,0 +1,117 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+#include "rgw_tracer.h"
+#include <yaml-cpp/yaml.h>
+
+namespace tracing {
+
+  const std::shared_ptr<opentracing::Tracer> Tracer::noop_tracer = opentracing::MakeNoopTracer();
+  TracerConfiguration tracing::Tracer::tracer_conf;
+
+  Tracer::Tracer(jaegertracing::Config& conf):open_tracer(jaegertracing::Tracer::make(conf)) {};
+
+  Tracer::Tracer(const std::string& config_file_path) {
+    auto config_yaml = YAML::LoadFile(config_file_path);
+    auto config = jaegertracing::Config::parse(config_yaml);
+    open_tracer =  jaegertracing::Tracer::make(config);
+  };
+
+  Tracer::~Tracer() {open_tracer->Close();}
+
+  std::unique_ptr<Span> Tracer::start_trace(const std::string& trace_name) {
+    std::unique_ptr<opentracing::Span> span;
+    if(is_enabled()) {
+      span = open_tracer->StartSpan(trace_name);
+    }
+    else {
+      span = noop_tracer->StartSpan(trace_name);
+    }
+    return std::unique_ptr<Span>(new Span(span));
+  }
+
+  std::unique_ptr<Span> Tracer::start_span(const std::string& span_name, std::unique_ptr<Span>& parent_span) {
+    std::unique_ptr<opentracing::Span> span;
+    if(is_enabled()) {
+      span = open_tracer->StartSpan(span_name, {parent_span->get_reference()});
+    }
+    else {
+      span = noop_tracer->StartSpan(span_name);
+    }
+    return std::unique_ptr<Span>(new Span(span));
```

Comment:
we shouldn't need to allocate a `Span` to wrap the `opentracing::Span` that's already heap-allocated. can we just get rid of this `Span` wrapper and use `opentracing::Span` directly?

Reply:
the reason is to call `Finish()` in the dtor (line 50)

Reply:
probably not needed. see: https://github.com/ceph/ceph/pull/42434/files#r677217884

Reply:
removed in ffd1262